materialize-firebolt/transactor.go

```diff
@@ -0,0 +1,287 @@
+	}
+
+	return jsonDoc, nil
+}
+
+// write file to S3 and keep a record of them to be copied to the table
+func (t *transactor) Store(it *pm.StoreIterator) error {
+	awsConfig := aws.Config{
+		Credentials: credentials.NewStaticCredentials(t.awsKeyId, t.awsSecretKey, ""),
+		Region:      &t.awsRegion,
+	}
+	sess := session.Must(session.NewSession(&awsConfig))
+	uploader := s3manager.NewUploader(sess)
+
+	uploads := []s3manager.BatchUploadObject{}
+
+	files := make(map[int][]byte)
+
+	for it.Next() {
+		doc, err := t.projectDocument(t.bindings[it.Binding].spec, it.Key, it.Values)
+		if err != nil {
+			return fmt.Errorf("projecting new store json: %w", err)
+		}
+
+		if _, ok := files[it.Binding]; ok {
+			files[it.Binding] = append(files[it.Binding], byte('\n'))
+			files[it.Binding] = append(files[it.Binding], doc...)
+		} else {
+			files[it.Binding] = doc
+		}
+	}
+
+	for bindingIndex, data := range files {
+		cp := t.checkpoint.Files[bindingIndex]
+
+		uploads = append(uploads, s3manager.BatchUploadObject{
+			Object: &s3manager.UploadInput{
+				Bucket: aws.String(cp.Bucket),
+				Key:    aws.String(cp.Key),
+				Body:   bytes.NewReader(data),
+			},
+		})
+	}
+
+	uploadIterator := s3manager.UploadObjectsIterator{
+		Objects: uploads,
+	}
+
+	err := uploader.UploadWithIterator(aws.BackgroundContext(), &uploadIterator)
+	if err != nil {
+		return fmt.Errorf("uploading files to s3: %w", err)
+	}
+
+	return nil
+}
+
+func (t *transactor) Commit(ctx context.Context) error {
+	return nil
+}
+
+// load stored file into table
+func (t *transactor) Acknowledge(context.Context) error {
+
+	if len(t.checkpoint.Files) < 1 {
+		return nil
+	}
+
+	sourceFileNames := map[int][]string{}
+	for _, item := range t.checkpoint.Files {
+		binding := item.Binding
+		fileName := item.Key
+
+		if _, ok := sourceFileNames[binding]; ok {
+			sourceFileNames[binding] = append(sourceFileNames[binding], fileName)
+		} else {
+			sourceFileNames[binding] = []string{fileName}
+		}
+	}
+
+	// Acknowledge step might end up being run by a separate process
+	// different from the original, and the new transactor process might be initialised with
+	// a new version of the MaterializationSpec. This means it's possible to have an Acknowledge
+	// with MaterializationSpec v1 be run by a transactor initialised with MaterializationSpec v2.
+	// As such, we keep a copy of the queries generated from the MaterializationSpec of the
+	// transaction to which this Acknowledge belongs in the checkpoint to avoid inconsistencies.
+	for binding, fileNames := range sourceFileNames {
+		insertQuery := t.checkpoint.Queries.Bindings[binding].InsertFromTable
+		values := fmt.Sprintf("'%s'", strings.Join(fileNames, "','"))
+		_, err := t.fb.Query(strings.Replace(insertQuery, "?", values, 1))
+		if err != nil {
+			return fmt.Errorf("moving files from external to main table: %w", err)
+		}
+	}
+
+	t.checkpoint.Files = []TemporaryFileRecord{}
+
+	return nil
+}
+
+func randomDocumentKey() string {
```

Comment:
nit: use `github.com/google/uuid.NewRandom()` ?

Reply:
That function has a possibility for error since it uses io operations, this one seems more straightforward without an error

Reply:
Mostly just reduction of SLOC and removing any possibility of a bug. And even if _we_ know there is no bug, it saves the reader time from wondering / verifying that as well. Being _very_ pedantic, I'd also be thinking about "hey, could these keys _ever_ collide?", and it's true that `rand` is not a cryptographically strong random number generator, while v4 UUID's are (drawing from `/dev/urandom`).

It technically returns an error (as `/dev/urandom` is file-like) but in your shoes I'd simply panic if one actually happens, as that's an infallible operation without something being seriously wrong.

Reply:
(This is a suggestion, not a requirement. You asked so I'm just unpacking the internal rationale that led to the suggestion).

Reply:
Makes sense 👍🏽